### ISSUE TITLE
output MMIF format 

### ISSUE THREAD

**Issue by keighrim** [clamsproject/role-filler-binder-new#1]
This thread to discuss output representation of R-F bindings in MMIF syntax and vocab.

**Comment by wricketts** [comment]
@keighrim  - I had some floating questions about RFB and general MMIF structure

1. When calling `mmif[<annotation_id>]` on a mmif object, I noticed that it only works if it's the annotation's `.long_id`. Using regular `.id` gives a KeyError. Is this intentional? 
1. Maybe I'm answering my own question, but I also noticed that the `.id` number is not unique globally, but only unique within a view. As in, `v_1` and `v_2`  can both have a TextDocument `td_1`  in them. Is there an implicit assumption that this `td_1` should correspond to the same document across views? If so, are there any innate enforcements/guards for that assumption, or are clams apps developers supposed to write logic complying with that assumption?

1. The RFB is implemented to return an empty csv if no roles/fillers are identified in the input (due to noise), or if the parser fails. @haydenmccormick brought up the suggestion that we add a runtime parameter to control whether or not the app should generate an annotation if the CSV content is empty. I thought this sounded reasonable, but had a concern related to the above 2 Q's. 

      If for example, docTR's `td_1` was too noisy, and the user opts to have RFB omit empty CSVs, then it's possible that RFB's `td_1`  could correspond to docTR's `td_2`  (or a higher number) , which is not super intuitive. Ultimately, the number mismatch won't prevent us from tracing the relation because we have `alignments`, but it could be less "user-friendly" to not have a global 1-to-1 mapping between id and document.

- Do we care about this?
- If we do, should we not include this runtime param, and have RFB always return annotations for each OCR textdocument?

**Comment by keighrim** [comment]
Regarding q1, I have started a new issue to make `id` unambiguous. https://github.com/clamsproject/mmif/issues/228 The problem is that when we start to force `long_id` everywhere, that'll break any future apps from past MMIFs (or past apps that generates past MMIFs). 

You are right about the annotation `id` without view-id prefix are implicitly "scoped" to the view it resides. That said, the annotation id can be re-used to refer to different objects as long as the "scope" is different. Thus, having `v1:td2` is aligned to `v2:td1`  is totally fine and we don't care. 

Al that put together, I don't think it's a good idea to produce "empty" text document when the RFB parsing fails - it doesn't add any information while adding space and time complexity to handle the MMIF outputs (storage-wise and `json.load`-wise).

**Comment by keighrim** [comment]
Revisiting this after a year; in the prototype implementation of the app we didn't really decide the MMIF-y representation of the RFB task outputs, namely 1) a set of roles 2) a set of fillers 3) a set of relations between them, and simply "dumped" the key-value structure as a tabular format (csv). At least, there was a suggestion to leave some special marker that the csv dump is not a regular `TextDocument` (https://github.com/clamsproject/clams-vocabulary/issues/11), but the followup actions after my original proposal was (sort of) rejected never took place and stalled for many months now. 

Plus, as the dumping to a csv format (or embedding already "structured" data within MMIF format in general) was a stop-gap solution anyway, I'd like to think about a real MMIF-y representation of the RFB task output. This will greatly help 
1. implement futher apps doing similar structural understading (notably a LLM-based app @kelleyl is working on) 
    - https://github.com/clamsproject/app-llava-captioner/blob/625aa4ba1a10ae180376ae49455388b19d108822/config/slates_all_fields.yaml
1. ongoing effort to implement a proper evaluation of such task    
    - https://github.com/clamsproject/aapb-evaluations/pull/66
    - https://github.com/clamsproject/aapb-annotations/issues/106

**Comment by keighrim** [comment]
One conceptually simple, but possibly technically complicated representation would be using [`NamedEntity`](http://vocab.lappsgrid.org/NamedEntity) types to wrap roles and fillers, then using a second "linking" type (e.g., [Relation](http://vocab.lappsgrid.org/Relation.html), [GenericRelation](http://vocab.lappsgrid.org/GenericRelation.html), [Alignment](https://mmif.clams.ai/vocabulary/Alignment/v1/) ) for relations. I say it's _possibly technically complicated_ because this will involved a long chain of back-referencing of annotation objects, for example, 

0. (app), (back-referencing annotation obejct), (reference mechanism) 
2. (source)`VideoDocument` (N/A)
3. (swt)`TimePoint` (time offset) 
4. (swt)`TimeFrame` (`representative` prop) 
5. (ocr) `TextDocument` (`Alignment` type)
6. (ocr) `Token` (char offsets) 
7. (rfb) `NamedEntity` (`targets` prop)  
8. (rfb) `Relation` (???)

**Comment by kelleyl** [comment]
This is an example text document from the current llava app. I was planning to change it to separate the prompt and response. If we were to take the approach you're suggesting, the app would also generate tokens, named entities - 2 for each key-value pair - one for the key and one for the value, with targets referencing the token ids, and a relation (?) linking the 2 named entities. Is that what you were thinking? 
```{
                    "@type": "http://mmif.clams.ai/vocabulary/TextDocument/v1",
                    "properties": {
                        "text": {
                            "@value": "[INST]  \nI have an image containing information about a TV show or movie.\n\nI need you to extract the following information from the image and generate it as JSON:\n\nprogram-title\nepisode-title\nseries-title\ntitle\nepisode-no\ncreate-date\nair-date\ndate\ndirector\nproducer\ncamera\n\nImportant Notes:\n- Not all of these keys will necessarily be present in the image.\n- The image may contain other information not listed above.\n- Only include the values for the keys listed above that are actually found in the image.\n- If a key is not found, do not include it in the JSON output.\n\nExample JSON Output:\n{\n  \"program-title\": \"The Simpsons\",\n  \"episode-title\": \"Homer Goes to College\",\n  \"episode-no\": \"5x02\" \n}\n\nPlease provide the JSON output based on the information extracted from the image. [/INST]\n\`\`\`json\n{\n  \"program-title\": \"STATE LINE\",\n  \"episode-title\": \"\",\n  \"series-title\": \"\",\n  \"title\": \"STATE LINE\",\n  \"episode-no\": \"705\",\n  \"create-date\": \"\",\n  \"air-date\": \"10/29/86 19:30\",\n  \"date\": \"\",\n  \"director\": \"HATCH\",\n  \"producer\": \"SHEEHAN\",\n  \"camera\": \"TH 2387\"\n}\n\`\`\`",
                            "@language": "en"
                        },
                        "id": "td_1"
                    }
                }```
```

**Comment by keighrim** [comment]
Basically, yes. But the example in my comment was assuming a modular workflow. With singleton, e2e approach, I don't think we need to generate all the intermediate types (for instance, `NamedEntity` can directly refer to a portion of the document using `start`/`end` props) .

**Comment by kelleyl** [comment]
That sounds good to me. My only initial concern was the number of intermediate annotations being generated, but since we can just refer to the original document with offsets instead of generating tokens that won't be an issue. This would definitely simplify the evaluation since we could just look for relations and what they are targeting instead of dealing with parsing whatever is in the text document.

**Comment by keighrim** [comment]
Other ideas from a whiteboard picture from a meeting between me and @marcverhagen about 1 month ago. 
- `Span`-based structure
    - `NamedEntity` (as described above)
    - `Span` or `Markable` (note that we are discussing dropping `markable` during lapps to clams vocab migration https://github.com/clamsproject/mmif/issues/202) in conjunction with some relation-based types (see below)
- `Relation`-based structure
    - generic `Relation` type with `head`, `arguments` properties
    - `KeyValue` (new) type with `key`, `value`, `role` properties
    - `Role` (new) type with `label`, `orga`, `per???` 
- non-consuming
    - `SemanticTag` type with properties like `target`, `label`, `role`.

(some part of the picture is blurry, hence leaving mystery prop names in the above. I will keep this comment updated as we go back to the whiteboard discussion.)

**Comment by keighrim** [comment]
Proposing a concrete representation based on the options discussed above (NamedEntity-based, Relation-based, SemanticTag-based from the whiteboard).

Proposal: `NamedEntity` + a new `Entity` type.

Full rationale and type spec at https://github.com/clamsproject/clams-vocabulary/issues/13.

The representation uses existing `NamedEntity` for role-labeled filler spans, and a new `Entity` type for record grouping and KB grounding. No new span types, no Relation subtypes, no SemanticTag (which was never a real LAPPS type).

Chyron example (`"John Smith\nCNN reporter"`):

```json
{
  "annotations": [
    {"@type": "NamedEntity", "id": "ne_1", "label": "name",        "start": 0,  "end": 10, "document": "v_ocr:td_1"},
    {"@type": "NamedEntity", "id": "ne_2", "label": "affiliation", "start": 11, "end": 14, "document": "v_ocr:td_1"},
    {"@type": "NamedEntity", "id": "ne_3", "label": "title",       "start": 15, "end": 23, "document": "v_ocr:td_1"},
    {"@type": "Entity",      "id": "ent_1", "...linked to...": ["ne_1", "ne_2", "ne_3"]}
  ]
}
```

(also see a slate example in https://github.com/clamsproject/clams-vocabulary/issues/13)

- `NamedEntity` is reused without modification: `NE.label` = role name from domain-specific labelset (e.g., `["name", "affiliation", "title"]` for chyrons, `["air-date", "director", "episode"]` for slates). NER uses `label` for entity types (`PERSON`, `ORG`) — different views, different labelsets, no conflict.
- `Entity` groups the NEs into one record via `origins` (proposed in https://github.com/clamsproject/clams-vocabulary/issues/12#issuecomment-3941639711 as a new base-level linking property, distinct from `targets`/`composition`)
- `Entity.uri` (optional) enables downstream NEL to ground the record or individual fillers to a KB (also unblocks https://github.com/clamsproject/app-dbpedia-spotlight-wrapper/issues/1)

For e2e apps (LLaVA, SmolVLM), the app produces a `TextDocument` first, then anchors NEs on it — same output structure regardless of whether OCR was a separate upstream step.

This replaces the "json-dump" stopgap and directly enables span-level evaluation (https://github.com/clamsproject/aapb-evaluations/pull/66) and pure MMIF-based structured visualization.

**Comment by keighrim** [comment]
adding a slate example 

`"AIR:2000/1/1\nDIR:John Smith\nEPI:#224"`

```json
{
  "annotations": [
    {"@type": "NamedEntity", "id": "ne_1", "label": "air-date",  "start":  4, "end": 12, "document": "v_ocr:td_1"},
    {"@type": "NamedEntity", "id": "ne_2", "label": "director",  "start": 17, "end": 27, "document": "v_ocr:td_1"},
    {"@type": "NamedEntity", "id": "ne_3", "label": "episode",   "start": 32, "end": 36, "document": "v_ocr:td_1"},
    {"@type": "Entity",      "id": "ent_1", "origins": ["ne_1", "ne_2", "ne_3"]}
  ]
}
```

**Comment by keighrim** [comment]
### Representing `name-normalized` in the NE+Entity model

In the current stopgap JSON-dump format used by the existing KIE apps (see https://github.com/clamsproject/app-heuristic-chyron-understanding/issues/1), we have a `name-normalized` field alongside `name-as-written` and `attributes`. With the NE+Entity representation proposed above, the as-written name is naturally captured by the NE span text, and attributes become additional NE spans. But the normalized form ("Smith, John") is a derived value with no home; it is not in the source text, and there is no KB to look it up from.

Note that with proposed `Entity` placed under the new `Meta` intermediate (rather than `Annotation`, see https://github.com/clamsproject/clams-vocabulary/issues/13#issuecomment-3980218256), it does not inherit the `label` property. And `label` in the CLAMS vocabulary is always classificatory (`NE.label="PERSON"`, `TF.label="chyron"`), not a display name. Hence, repurposing it for the normalized name would be semantically wrong anyway.

#### Option A: compositional approach using existing primitives

One option is to avoid adding a new property and instead use existing vocabulary types compositionally: create a new TD containing the normalized name value (with `origins` pointing to the original name NE), then anchor a new `NE(label="name")` on that derived TD, and finally have the Entity originate from all relevant NEs, both the original spans (name, affiliation, title, ...) and the normalized-name NE:

```
# KIE view (same as before)
NE(id="ne_1", label="name",        start=0,  end=10, document="td_1")  # "John Smith"
NE(id="ne_2", label="affiliation", start=11, end=14, document="td_1")  # "CNN"
NE(id="ne_3", label="title",       start=15, end=23, document="td_1")  # "reporter"

# Still KIE app, additionally performing name normalization (same view or separate)
TD(id="td_norm", text="Smith, John", origins=["ne_1"])
NE(id="ne_4", label="name", start=0, end=11, document="td_norm")

# Entity grouping
Entity(id="ent_1", origins=["ne_1", "ne_2", "ne_3", "ne_4"])
```

This avoids adding a new machinery but the representational overhead is significant; an extra TD + NE per normalized name, just to store a simple string transformation. It also muddies the Entity's `origins`: `ne_1` and `ne_4` both carry `label="name"` but refer to different texts on different documents, and a consumer has no way to know which is the "canonical" one without inspecting the TD derivation chain (we can rename the label in the normalized case, but that requires an additional controlled vocabulary).

#### Option B: `normalizedName` property on `Entity` (mine preferred)

An optional `normalizedName` (working name) property on `Entity` (not on `Meta`; Documents don't need display names). It stores the canonical human-readable form of the entity's name, migrating the `name-normalized` field from the JSON-dump stopgap directly into the proper vocabulary.

`normalizedName` and `location` (the URI property being promoted to `Meta`) are effectively **mutually exclusive**: when an entity is grounded via `location` (e.g., `location="wd:Q12345"`), the canonical name should be served by the KB entry behind that URI: **it should not be leaked into the MMIF representation**. `normalizedName` is specifically for underspecified entities where no KB grounding is available.

Name normalization is reliable today (regex- or LLM-based), while full NEL (linking to Wikidata, LoC, etc.) is not. `normalizedName` gives KIE apps a place to store normalized names without requiring a grounding step that may not be available.

I strongly prefer Option B. At the end of the day, `normalizedName` is just a poor man's semi-grounding — a best-effort canonical label for entities that don't have a KB URI yet. It's not worth the complexity of the whole Option A circus (minting a TD, anchoring an NE, threading it through `origins`) just to store a single derived string. A dedicated property is simpler for producers, simpler for consumers, and directly mirrors the field we already have in the JSON-dump stopgap. That said, I'd prefer to keep `normalizedName` as just an `additionalProperty` convention rather than an official controlled vocabulary property. Alternatively, we could declare it in the vocab definition as an optional property. But given how narrow the use case is, I'm not sure it warrants that level of formality.

#### Either way

The normalization rule itself (e.g., "last, first" reordering) is a property of the producing app, should be recorded in view/app metadata; it does not need to be specified on the Entity object.
